Package XLA FFI headers with jaxlib wheel

dfm opened the pull request:
The new "typed" API that XLA provides for foreign function calls is header-only and packaging it as part of jaxlib could simplify the open source workflow for building custom calls.

It's not completely obvious that we need to include this, because jaxlib isn't strictly required as a _build_ dependency for FFI calls, although it typically will be required as a _run time_ dependency. Also, it probably wouldn't be too painful for external projects to use the headers directly from the openxla/xla repo.

All that being said, I wanted to figure out how to do this, and it has been requested a few times.

dfm committed:
Package XLA FFI headers with jaxlib wheel

The new "typed" API that XLA provides for foreign function calls is
header-only and packaging it as part of jaxlib could simplify the open
source workflow for building custom calls.

It's not completely obvious that we need to include this, because jaxlib
isn't strictly required as a _build_ dependency for FFI calls, although
it typically will be required as a _run time_ dependency. Also, it
probably wouldn't be too painful for external projects to use the
headers directly from the openxla/xla repo.

All that being said, I wanted to figure out how to do this, and it has
been requested a few times.

superbobry commented on jaxlib/setup.py:
```diff
@@ -118,6 +118,7 @@ def has_ext_modules(self):
             'triton/*.pyi',
             'triton/*.pyd',
             'triton/*.so',
+            'include/xla/ffi/api/*.h',
```
Yeah, I think this should only include the files copied by build_wheel.py.

superbobry reviewed (approved):
I assume you built the wheel locally and verified that the header files are included? If not, I highly recommend doing that before submitting :)

dfm commented on jaxlib/init.py:
```diff
@@ -12,4 +12,9 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+def include_dir() -> str:
```
Good point! Perhaps others can comment, but I don't see a great way to test this, but I do have it on my to do list to add a simple tutorial demonstrating the use.

andportnoy replied:
Maybe something as simple as
```python
import jaxlib
jaxlib.include_dir()
```
?
Just to check that the function exists and is accessible.

hawkinsp commented on jaxlib/init.py:
```diff
@@ -12,4 +12,9 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+def include_dir() -> str:
```
I think we should put this Python API in *jax*. In general jaxlib has no public APIs and I'd like to keep it that way.

This function is "where do I find the headers" and it can point to jaxlib. But it should be a jax API so we can change the package structure later.

dfm commented:
I've incorporated the suggestions from @hawkinsp and @andportnoy: I moved the `include_dir` interface into `jax`, which also makes it more straightforward to test. I'm inclined to defer documenting this feature to be included in an update to the custom call tutorials, unless anyone feels strongly otherwise!

hawkinsp commented on jax/__init__.py:
```diff
@@ -126,6 +126,7 @@
 from jax._src.api import vmap as vmap
 from jax._src.api import xla_computation as xla_computation
 from jax._src.sharding_impls import NamedSharding as NamedSharding
+from jax._src.include_dir import include_dir as include_dir
```
I'm wondering if we should use a new submodule for this, perhaps a submodule related to defining foreign functions.

dfm replied:
I think that's a good suggestion! I'll update the PR to have this function in `jax.ffi` perhaps. I don't think that there's currently have anything else that would go in there, but I think that it clarifies the use case.

Another option would be to call the submodule something like `custom_call`, since I expect that we will be adding more infrastructure to support custom calls, and that would be the main interface to the FFI.

dfm commented:
Thanks all for your feedback! I think this is good to go, but I just wanted to double check @hawkinsp's thoughts on the final layout. Let me know if you have any other comments!
